api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/AbstractQueryBuilder.java

```diff
@@ -1,27 +1,186 @@
+import java.util.stream.Collectors;
 
 /**
  * AbstractQueryBuilder is an object that builds {@link QueryJobConfiguration}
  * for BigQuery.
  */
 public abstract class AbstractQueryBuilder {
 
-    /**
-     * Build a {@link QueryJobConfiguration} from the specified
-     * {@link QueryParameters} provided.
-     *
-     * @param parameters
-     * @return
-     */
-    public abstract QueryJobConfiguration buildQueryJobConfig(QueryParameters parameters);
+  ImmutableMap<ModifierType, String> exceptionText = ImmutableMap.<ModifierType, String>builder()
+    .put(ModifierType.AGE_AT_EVENT, "age at event")
+    .put(ModifierType.EVENT_DATE, "event date")
+    .put(ModifierType.NUM_OF_OCCURRENCES, "number of occurrences")
+    .build();
 
-    public abstract FactoryKey getType();
+  public static final String AGE_AT_EVENT_PREFIX = "age";
+  public static final String EVENT_DATE_PREFIX = "event";
+  public static final String OCCURRENCES_PREFIX = "occ";
 
-    protected String getUniqueNamedParameterPostfix() {
-        return UUID.randomUUID().toString().replaceAll("-", "");
+  public static final String WHERE = " where ";
+  public static final String AND = " and ";
+  private static final String MODIFIER_SQL_TEMPLATE = "select criteria.person_id from (${innerSql}) criteria\n";
+  private static final String AGE_AT_EVENT_JOIN_TEMPLATE =
+    "join `${projectId}.${dataSetId}.person` p on (criteria.person_id = p.person_id)\n";
+  private static final String AGE_AT_EVENT_SQL_TEMPLATE =
+    "CAST(FLOOR(DATE_DIFF(criteria.entry_date, DATE(p.year_of_birth, p.month_of_birth, p.day_of_birth), MONTH)/12) as INT64)\n";
+  private static final String EVENT_DATE_SQL_TEMPLATE = "criteria.entry_date\n";
+  private static final String OCCURRENCES_SQL_TEMPLATE = "group by criteria.person_id\n" +
+    "having count(criteria.person_id)";
+
+  /**
+   * Build a {@link QueryJobConfiguration} from the specified
+   * {@link QueryParameters} provided.
+   *
+   * @param parameters
+   * @return
+   */
+  public abstract QueryJobConfiguration buildQueryJobConfig(QueryParameters parameters);
+
+  public QueryJobConfiguration buildModifierSql(String baseSql, Map<String, QueryParameterValue> queryParams, List<Modifier> modifiers) {
+    String modifierSql = "";
+    String finalSql = "";
+    if (!modifiers.isEmpty()) {
+      Modifier ageAtEvent = getModifier(modifiers, ModifierType.AGE_AT_EVENT);
+      Modifier eventDate = getModifier(modifiers, ModifierType.EVENT_DATE);
+      Modifier occurrences = getModifier(modifiers, ModifierType.NUM_OF_OCCURRENCES);
+      modifierSql = buildAgeAtEventAndEventDateModifierSql(queryParams, Arrays.asList(ageAtEvent, eventDate));
+      //Number of Occurrences has to be last because of the group by
+      modifierSql = modifierSql + buildNumOfOccurrencesModifierSql(queryParams, occurrences);
+    }
+    finalSql = MODIFIER_SQL_TEMPLATE.replace("${innerSql}", baseSql) +
+      modifierSql;
+
+    return QueryJobConfiguration
```

Comment:
Is it possible to just return the modifiersql string, and put together this finalsql builder in the specific criteria, for consistency with other non-modifier criteria?

Reply:
Yes, I will change. Done.